feat: Add support for creating autolink references

hoffm opened the pull request:
Fixes #10119.

## Overview

Adds support for `gh repo autolink create` command. The command requires two positional arguments for the key prefix and URL template, respectively. It provides one custom flag `--numeric`/`-n` to specify a non-alphanumeric autolink.

When creation is successful, I took the liberty of adding some information to the output in excess of what's in the spec (see examples below). The output includes the ID of the created autolink and also a representation of the mapping (prefix to URL) it represents. The mapping is intended to be similar to what the github.com UI provides.

Non-201 API responses are reported transparently from the response via [HandleHTTPError](https://pkg.go.dev/github.com/cli/cli/v2/api#HandleHTTPError). The one exception is 404 responses, for which we rewrite the message to report "Must have admin rights to Repository" as specified.

## Examples

### Help
![create go — github-cli 2025-01-05 at 8 34 23 PM jpg](https://github.com/user-attachments/assets/d6fe3deb-e6f0-4ab9-af27-fb86505267d9)

### Create

![http_test go — github-cli 2025-01-05 at 3 48 51 PM jpg](https://github.com/user-attachments/assets/f37def23-c6c3-43a8-84c5-f63ae8a2776b)

### Create Numeric

![http_test go — github-cli 2025-01-05 at 3 50 21 PM jpg](https://github.com/user-attachments/assets/c326ad32-5db8-409d-8d92-ae052dd196f0)

### Key Prefix Already Exists

![http_test go — github-cli 2025-01-05 at 3 49 25 PM](https://github.com/user-attachments/assets/2b7e8110-ed32-4d5e-a954-c18c3e146507)

### Non-Admin (404)

![http_test go — github-cli 2025-01-05 at 3 54 58 PM jpg](https://github.com/user-attachments/assets/771e7648-cb19-456c-8098-9f69bbfe0098)

### URL Template missing `<num>`

![http_test go — github-cli 2025-01-05 at 3 55 31 PM jpg](https://github.com/user-attachments/assets/7f578ad7-5148-4e1d-b52c-966d55664c5c)

### URL Template not URL

![http_test go — github-cli 2025-01-05 at 3 56 08 PM jpg](https://github.com/user-attachments/assets/a5389a4f-e3a6-49ff-9880-b540308bfa0c)

### Multiple Field Validation Failures

![http_test go — github-cli 2025-01-05 at 3 54 13 PM jpg](https://github.com/user-attachments/assets/bb2fb1fb-1d48-4c8c-b699-478dba80ea6a)

## Software Design Considerations

I've created a new `shared` package to house the `Autolink` data model, which needs to be shared between `create`, `list`, and future command packages. To maintain the current design, this type needs to be in a new package to avoid import cycles between `autolink` and the command packages. I'm open to renaming this (or reconsidering the design). 

The "shared" package is at heart is also a step towards domain-driven design. If we wanted to go a step further towards DDD, we could think of the REST API as part of a persistence layer that implements an `AutolinkClient` interface defined in the domain:

```go
type AutolinkClient interface {
    Create(repo ghrepo.Interface, request AutolinkCreateRequest) (*domain.Autolink, error)
    List(repo ghrepo.Interface) ([]domain.Autolink, error)
    // etc
}
```

Doing this would require quite a different directory and package structure than we have today, so I didn't want to rush into it.

## Questions

1. I edited the the `RESTPayload` helper in [pkg/httpmock/stub.go](https://github.com/cli/cli/pull/10180/files#diff-30f7a09ce6fb5698f7099daca69b7d6e40b8b58906f272d590b3d4552310c5c5). This helper was not setting a content type header on the stubbed response, and as a result it was incompatible with [HandleHTTPError](https://pkg.go.dev/github.com/cli/cli/v2/api#HandleHTTPError). My change sets this header to `"application/json"`, which matches the behavior of `JSONResponse` in the same package. Is that cool?
2. I noticed that on my personal repos I am able to create autolink references via the API, but the settings page doesn't have a link to the GUI to view them. When I visit the URL that should have this GUI, `https://github.com/nytimes/OWNER/REPO/key_links`, I get a 404. If this feature is not available on all repos, should that be reflected in the responses provided by this tool, or by the REST API itself?

hoffm committed:
feat: Add support for creating autolink references

hoffm committed:
Merge branch 'trunk' into create-autolinks

hoffm committed:
Fix typos

hoffm committed:
Merge branch 'cli:trunk' into create-autolinks

hoffm committed:
Merge branch 'trunk' into create-autolinks

hoffm committed:
Merge branch 'trunk' into create-autolinks

hoffm committed:
Merge branch 'trunk' into create-autolinks

hoffm committed:
Merge branch 'trunk' into create-autolinks

hoffm committed:
Merge branch 'trunk' into create-autolinks

hoffm committed:
Merge branch 'trunk' into create-autolinks

hoffm commented:
@andyfeller Could you help me get code review on this PR? I've started work on the other issues with the same parent ticket, but I won't be able to finish anything else until we resolve this one.

williammartin commented:
It's on my ever expanding todo list, sorry! But I'm happy for someone else to give it a review as well.

I'd prefer `domain` to be called `shared` to be consistent with other packages around the codebase. To be honest I'd prefer the command to go under a `cmd` or `cobra` package and to use the `autolink` package as the domain package itself, but that's too much of a departure from the current state right now.

hoffm commented:
> It's on my ever expanding todo list, sorry! But I'm happy for someone else to give it a review as well.
> 
> I'd prefer `domain` to be called `shared` to be consistent with other packages around the codebase. To be honest I'd prefer the command to go under a `cmd` or `cobra` package and to use the `autolink` package as the domain package itself, but that's too much of a departure from the current state right now.

Sorry to nag, @williammartin! It's truly a reflection of my excitement to keep going, not of any annoyance with you or the other maintainers. :)

Roger that re: "shared".

hoffm committed:
Rename domain pkg to shared

jtmcg commented:
Given I reviewed #10124 I'm happy to give this a look, @williammartin and @hoffm. I also have a few other priorities ahead of this right now, so we'll see who gets to it first 🙂

hoffm committed:
Merge branch 'trunk' into create-autolinks

hoffm committed:
Merge branch 'trunk' into create-autolinks

hoffm committed:
Merge branch 'trunk' into create-autolinks

hoffm commented:
Hey team. It looks like I'll have some bandwidth to work on this project this week. If it's possible to get a review here, I'd appreciate it!

andyfeller commented:
> Hey team. It looks like I'll have some bandwidth to work on this project this week. If it's possible to get a review here, I'd appreciate it!

@hoffm : apologies for the delay on this; I'm going to review this today.

I'll note there are a few minor format things that I might just make to the branch to avoid some of the back and forth; nothing major.

andyfeller committed:
Minor refactoring autolink create help and logic

- simplified and wrapped `gh repo autolink create` and `gh repo autolink` long help usage docs
- simplified success message, brought into alignment with other commands

andyfeller commented:
> > Hey team. It looks like I'll have some bandwidth to work on this project this week. If it's possible to get a review here, I'd appreciate it!
> 
> @hoffm : apologies for the delay on this; I'm going to review this today.
> 
> I'll note there are a few minor format things that I might just make to the branch to avoid some of the back and forth; nothing major.

@hoffm : I finished testing and pushing up https://github.com/cli/cli/pull/10180/commits/7c31d1a76b6c7a0e3eeb71ccb34b55b679e5fef5 for some minor changes and think this is in a good place.  Let me explain my changes:

1. Some of the long help usage content in `gh repo autolink create` felt like it should be in the commandset help doc; specifically that `admin` role is required to use any of the commands
2. The really long help usage lines needed to be broken into digestible blocks
3. I know I originally drafted the AC regarding `--numeric` flag with that language, but in retrospect something @williammartin said made me think that changing that language a little was worthwhile
4. I discovered a minor test assertion mix up of actual vs expected
5. I brought the success message more in line with [other messages in `gh`](https://github.com/search?q=repo%3Acli%2Fcli%20created%20%22fmt.Fprintf%22&type=code)

Again, thank you for your time supporting `gh` and patience as we deal with a bit of chaos at the moment! 🙇 

_Side note: I know the following error comes from the REST API but I don't like it_ 😜 🤷 

```shell
$ ~/.../cli/cli/bin/gh repo autolink create
Cannot create autolink: keyPrefix and urlTemplate arguments are both required

Usage:  gh repo autolink create <keyPrefix> <urlTemplate> [flags]

Flags:
  -n, --numeric   Mark autolink as numeric
  
$ ~/.../cli/cli/bin/gh repo autolink create ABC "http://example.com"
HTTP 422: Validation Failed (https://api.github.com/repos/tinyfists/gh-nonsense/autolinks)
url_template is missing a <num> token


$ ~/.../cli/cli/bin/gh repo autolink create ABC "http://example.com?q=<num>"
✓ Created repository autolink 7160931 on tinyfists/gh-nonsense


$ ~/.../cli/cli/bin/gh repo autolink list                                   

Showing 1 autolink reference in tinyfists/gh-nonsense

ID       KEY PREFIX  URL TEMPLATE                ALPHANUMERIC
7160931  ABC         http://example.com?q=<num>  true
```

andyfeller reviewed (approved):
LGTM

hoffm commented:
@andyfeller Your changes all make sense to me. I'll take account of these patterns in future work. Thanks!
